test/test_buffer_protocol.py

```diff
@@ -0,0 +1,175 @@
+import torch.testing._internal.common_utils as common
+from torch.testing._internal.common_device_type import (
+    instantiate_device_type_tests,
+    dtypes, onlyCPU
+)
+
+import torch
+import numpy
+
+def get_dtype_size(dtype):
+    return int(torch.empty((), dtype=dtype).element_size())
+
+SIZE = 5
+SHAPE = (SIZE,)
+
+# Tests for the `frombuffer` function (only work on CPU):
+#   Constructs tensors from Python objects that implement the buffer protocol,
+#   without copying data.
+class TestBufferProtocol(common.TestCase):
+    def _run_test(self, shape, dtype, count=-1, first=0, offset=None, **kwargs):
+        numpy_dtype = common.torch_to_numpy_dtype_dict[dtype]
+
+        if offset is None:
+            offset = first * get_dtype_size(dtype)
+
+        numpy_original = common.make_tensor(shape, torch.device("cpu"), dtype).numpy()
+        original = memoryview(numpy_original)
+        # First call PyTorch's version in case of errors.
+        # If this call exits successfully, the NumPy version must also do so.
+        torch_frombuffer = torch.frombuffer(original, dtype=dtype, count=count, offset=offset, **kwargs)
+        numpy_frombuffer = numpy.frombuffer(original, dtype=numpy_dtype, count=count, offset=offset)
+
+        self.assertEqual(numpy_frombuffer, torch_frombuffer)
+        self.assertEqual(numpy_frombuffer.__array_interface__["data"][0], torch_frombuffer.data_ptr())
+        return (numpy_original, torch_frombuffer)
+
+    @onlyCPU
+    @dtypes(*common.torch_to_numpy_dtype_dict.keys())
+    def test_same_type(self, device, dtype):
+        self._run_test((), dtype)
+        self._run_test((4,), dtype)
+        self._run_test((10, 10), dtype)
+
+    @onlyCPU
+    @dtypes(*common.torch_to_numpy_dtype_dict.keys())
+    def test_requires_grad(self, device, dtype):
+        def _run_test_and_check_grad(requires_grad, *args, **kwargs):
+            kwargs["requires_grad"] = requires_grad
+            _, tensor = self._run_test(*args, **kwargs)
+            self.assertTrue(tensor.requires_grad == requires_grad)
+
+        if dtype.is_floating_point or dtype.is_complex:
```

Comment:
Nit: this could also be modeled like this:

```
requires_grad = dtype.is_floating_point or dtype.is_complex
_run_test_and_check_grad(requires_grad, (), dtype)
_run_test_and_check_grad(requires_grad, (4,), dtype)
_run_test_and_check_grad(requires_grad, (10, 10), dtype)
```

Reply:
Ah, very nice observation. Thanks.